presto-iceberg/src/main/java/com/facebook/presto/iceberg/rest/IcebergRestCatalogFactory.java

```diff
@@ -77,22 +126,99 @@ protected Map<String, String> getCatalogProperties(ConnectorSession session)
 
         catalogConfig.getSessionType().ifPresent(type -> {
             if (type == USER) {
-                properties.putAll(session.getIdentity().getExtraCredentials());
+                properties.put(CatalogProperties.USER, session.getUser());
+            }
+        });
 
+        return properties.build();
+    }
+
+    protected SessionContext convertSession(ConnectorSession session)
+    {
+        RestSessionBuilder sessionContextBuilder = builder(session);
+        catalogConfig.getSessionType().ifPresent(type -> {
+            if (type.equals(USER)) {
                 String sessionId = format("%s-%s", session.getUser(), session.getSource().orElse("default"));
+
+                Map<String, String> properties = ImmutableMap.of(
+                        "user", session.getUser(),
+                        "source", session.getSource().orElse(""),
+                        "version", nodeVersion.toString());
+
                 String jwt = Jwts.builder()
-                        .setId(sessionId)
                         .setSubject(session.getUser())
-                        .setIssuedAt(new Date())
                         .setIssuer(nodeVersion.toString())
+                        .setIssuedAt(new Date())
                         .claim("user", session.getUser())
                         .claim("source", session.getSource().orElse(""))
                         .compact();
 
-                properties.put(OAuth2Properties.JWT_TOKEN_TYPE, jwt);
+                ImmutableMap.Builder<String, String> credentials = ImmutableMap.builder();
+                credentials.put(JWT_TOKEN_TYPE, jwt).putAll(session.getIdentity().getExtraCredentials());
+
+                sessionContextBuilder.setSessionId(sessionId)
+                        .setIdentity(session.getUser())
+                        .setCredentials(credentials.build())
+                        .setProperties(properties);
             }
         });
+        return sessionContextBuilder.build();
+    }
 
-        return properties.build();
+    protected static class RestSessionBuilder
+    {
+        private String sessionId;
+        private String identity;
+        private Map<String, String> properties;
+        private Map<String, String> credentials;
+        private final ConnectorIdentity wrappedIdentity;
+
+        private RestSessionBuilder(ConnectorSession session)
+        {
+            sessionId = randomUUID().toString();
```

Comment:
Wondering if we should default to using the query ID instead here. Though I do notice we never actually leave the default in place. The only caller of `setSessionId` always uses the pattern `$user-$source`. Do you think using query ID might be easier to trace issues in the future? Or was there a specific reason to use the `$user-$source` pattern?

Reply:
This raises a few good points. As far as the default session ID, it does remain in place when the REST session type is `NONE` (which is the default config value and likely the more common situation at the moment). That being said, perhaps I should be more explicit about setting it in an `orElse` block in the `convertSession` function for clarity.

I'm not tied to using a random UUID for the session ID here, but I'm also not sure query ID would have the desired affect either. Essentially, we want REST user sessions to persist across queries (again, assuming `SessionType` is set to `USER`). This is where the session Id comes in - it's purpose is for server-side state caching. I figured the combo of 'user' and 'source' would be the most useful for persisting state. That being said, I'm not very well-versed in server-side handling of REST sessions in general so I'd be happy to change this if you think there's a better way. Perhaps paring it down to just user would work.

All this to say that session Id is somewhat useless if user sessions aren't active, hence the use of a random UUID. I don't think it would help much for troubleshooting either way because session Id will really only be set when the catalog is created (or, more generally, on cache miss).

Let me know your thoughts - will push a fix for your other review comment shortly as well.

Reply:
I think we can leave it as a random UUID. I was curious about the reasoning behind it. I also didn't realize this wouldn't be set for non-`USER` session types. I think it's fine to leave as is. Thanks for the explanation!